data-prepper-plugins/key-value-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java

```diff
@@ -179,79 +240,215 @@ private void validateKeySets(final Set<String> includeSet, final Set<String> exc
-
-                if (!includeKeysSet.isEmpty() && !includeKeysSet.contains(key)) {
-                    LOG.debug("Skipping not included key: '{}'", key);
-                    continue;
+            if (keyValueProcessorConfig.getRecursive()) {
+                ObjectMapper mapper = new ObjectMapper();
+                try {
+                    JsonNode recursedTree = recurse(groupsRaw, mapper);
+                    outputMap.putAll(createRecursedMap(recursedTree, mapper));
+                } catch (Exception e) {
+                    LOG.error("Recursive parsing ran into an unexpected error, treating message as non-recursive");
                 }
+            } else {
+                outputMap.putAll(createNonRecursedMap(groups));
+            }
 
-                if (excludeKeysSet.contains(key)) {
-                    LOG.debug("Key is being excluded: '{}'", key);
-                    continue;
-                }
+            final Map<String, Object> processedMap = executeConfigs(outputMap);
 
-                if(keyValueProcessorConfig.getDeleteKeyRegex() != null && !Objects.equals(keyValueProcessorConfig.getDeleteKeyRegex(), "")) {
-                    key = key.replaceAll(keyValueProcessorConfig.getDeleteKeyRegex(), "");
-                }
-                key = keyValueProcessorConfig.getPrefix() + key;
+            recordEvent.put(keyValueProcessorConfig.getDestination(), processedMap);
+        }
 
-                if (terms.length == 2) {
-                    value = terms[1];
-                } else {
-                    LOG.debug("Unsuccessful match: '{}'", terms[0]);
-                    value = keyValueProcessorConfig.getNonMatchValue();
-                }
+        return records;
+    }
+
+    private ObjectNode recurse(String input, ObjectMapper mapper) {
+        Stack<Character> bracketStack = new Stack<Character>();
+        Map<Character, Character> bracketMap = initBracketMap();
+        int pairStart = 0;
 
-                if(value != null
-                        && value instanceof String
-                        && keyValueProcessorConfig.getDeleteValueRegex() != null
-                        && !Objects.equals(keyValueProcessorConfig.getDeleteValueRegex(), "")) {
-                    value = ((String)value).replaceAll(keyValueProcessorConfig.getDeleteValueRegex(), "");
+        ArrayList<String> pairs = new ArrayList<String>();
+        ObjectNode root = mapper.createObjectNode();
+
+        for (int i = 0; i < input.length(); i++) {
+            if (bracketMap.containsKey(input.charAt(i))) {
+                bracketStack.push(input.charAt(i));
+            }
+
+            if (bracketMap.containsValue(input.charAt(i)) && !bracketStack.isEmpty()) {
+                if (bracketMap.get(bracketStack.peek()) == input.charAt(i)) {
+                    bracketStack.pop();
                 }
+            }
+
+            if (bracketStack.isEmpty() && input.charAt(i) == fieldDelimiterPattern.toString().charAt(0)) {
```

Comment:
We should still use regex match even if the delimiter is restricted to a single character, i.e.
```
fieldDelimiterPattern.matcher(input.charAt(i)).matches()
```
Here fieldDelimiterPattern is a `Pattern` from either field_delimiter_regex or field_delimiter_character